src/dragonball/src/resource_manager.rs

```diff
@@ -0,0 +1,683 @@
+// Copyright (C) 2022 Alibaba Cloud. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Mutex;
+
+use dbs_allocator::{Constraint, IntervalTree, Range};
+use dbs_boot::layout::{
+    GUEST_MEM_END, GUEST_MEM_START, GUEST_PHYS_END, IRQ_BASE as LEGACY_IRQ_BASE,
+    IRQ_MAX as LEGACY_IRQ_MAX, MMIO_LOW_END, MMIO_LOW_START,
+};
+use dbs_device::resources::{DeviceResources, MsiIrqType, Resource, ResourceConstraint};
+
+// The LEGACY_IRQ_BASE irq is reserved for shared IRQ.
+const SHARED_IRQ: u32 = LEGACY_IRQ_BASE;
+#[cfg(target_arch = "x86_64")]
+const MSI_IRQ_BASE: u32 = 24;
+#[cfg(target_arch = "aarch64")]
+/// It's gsi base id of msi
+const MSI_IRQ_BASE: u32 = LEGACY_IRQ_MAX + 1;
+
+// Currently, The max number of MSI IRQ we support is 1023
+const MSI_IRQ_MAX: u32 = 1023;
+#[cfg(target_arch = "x86_64")]
+const KVM_USER_MEM_SLOTS: u32 = 509;
+#[cfg(target_arch = "aarch64")]
+const KVM_USER_MEM_SLOTS: u32 = 512;
+const PIO_MIN: u16 = 0x0;
+const PIO_MAX: u16 = 0xFFFF;
+
+/// Errors associated with resource management operations
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum ResourceError {
+    /// Unknown/unsupported resource type.
+    #[error("unsupported resource type")]
+    UnknownResourceType,
+
+    /// Invalid resource range.
+    #[error("invalid resource range")]
+    InvalidResourceRange,
+
+    /// No source available.
+    #[error("no resource available")]
+    NoAvailResource,
+}
+
+struct ResoureManagerBuilder {
+    legacy_irq_pool: IntervalTree<()>,
+    msi_irq_pool: IntervalTree<()>,
+    pio_pool: IntervalTree<()>,
+    mmio_pool: IntervalTree<()>,
+    mem_pool: IntervalTree<()>,
+    kvm_mem_slot_pool: IntervalTree<()>,
+}
+
+impl Default for ResoureManagerBuilder {
+    fn default() -> Self {
+        ResoureManagerBuilder {
+            legacy_irq_pool: IntervalTree::new(),
+            msi_irq_pool: IntervalTree::new(),
+            pio_pool: IntervalTree::new(),
+            mmio_pool: IntervalTree::new(),
+            mem_pool: IntervalTree::new(),
+            kvm_mem_slot_pool: IntervalTree::new(),
+        }
```

Comment:
`IntervalTree` provides `Default`, so this could become `ResoureManagerBuilder::default()` if you add `#[Default]` to `ResoureManagerBuilder` I think?

Reply:
changed